src/stan/callbacks/concurrent_writer.hpp

```diff
@@ -0,0 +1,121 @@
+ * parallel for loop, and all threads hit an instance of max capacity. TBB can
+ * choose to wait for a thread to finish instead of spinning up the write
+ * thread. So to circumvent that issue, we use an std::thread.
+ * @tparam Writer A type that inherits from `writer`
+ */
+template <typename Writer>
+struct concurrent_writer {
+  // A reference to the writer to write to
+  std::reference_wrapper<Writer> writer;
+  // Queue for string messages
+  tbb::concurrent_bounded_queue<std::string> str_messages_{};
+  // Queue for vector of strings messages
+  tbb::concurrent_bounded_queue<std::vector<std::string>> vec_str_messages_{};
+  // Queue for Eigen vector messages
+  tbb::concurrent_bounded_queue<Eigen::RowVectorXd> eigen_messages_{};
+  // Flag to stop the writing thread once all queues are empty
+  bool continue_writing_{true};
+  // The writing thread
+  std::thread thread_;
+  /**
+   * Constructs a concurrent writer from a writer and spins up a thread for
+   * writing.
+   * @param writer A writer to write to
+   */
+  explicit concurrent_writer(Writer& writer) : writer(writer) {
+    str_messages_.set_capacity(1000);
+    vec_str_messages_.set_capacity(1000);
+    eigen_messages_.set_capacity(1000);
+    thread_ = std::thread([&]() {
+      std::string str;
+      std::vector<std::string> vec_str;
+      Eigen::RowVectorXd eigen;
+      while (continue_writing_
+             || !(str_messages_.empty() && vec_str_messages_.empty()
+                  && eigen_messages_.empty())) {
+        while (str_messages_.try_pop(str)) {
+          writer(str);
+        }
+        while (vec_str_messages_.try_pop(vec_str)) {
+          writer(vec_str);
+        }
+        while (eigen_messages_.try_pop(eigen)) {
+          writer(eigen);
+        }
+      }
+    });
+  }
+  /**
+   * Place a value in a queue for writing.
+   * @tparam T Either an `std::vector<std::string|double>`, an Eigen vector, or
+   * a string
+   * @param t A value to put on a queue
+   */
+  template <typename T>
+  void operator()(T&& t) {
+    if constexpr (stan::is_std_vector<T>::value) {
+      if constexpr (std::is_arithmetic_v<stan::value_type_t<T>>) {
+        eigen_messages_.push(Eigen::RowVectorXd::Map(t.data(), t.size()));
+      } else {
+        vec_str_messages_.push(t);
+      }
+    } else if constexpr (std::is_same_v<T, std::string>) {
+      str_messages_.push(std::forward<T>(t));
+    } else if constexpr (stan::is_eigen_row_vector<T>::value
+                         || stan::is_eigen_col_vector<T>::value) {
+      eigen_messages_.push(std::forward<T>(t));
+    } else {
+      static_assert(1,
+                    "Unsupported type passed to concurrent_writer. "
+                    "This is an internal error. Please file an issue on the "
+                    "stan github repo.");
+    }
+  }
+  /**
+   * Writes a comment prefix to the writer.
+   */
+  void operator()() { str_messages_.push(writer.get().comment_prefix()); }
```

Comment:
Oh I see, the `comment_prefix` exists so that this empty overload is implementable.

I think it would be better to keep that detail inside this class and have it contain something like an `std::atomic<int> null_writes_queued` which is incremented in this and decremented by the writer thread.


It also seems like this overload would primarily exist to control the formatting of the result, and anything calling this in a threaded context is likely gonna not get what it wants. Have you investigated why this overload gets called by pathfinder and what it is hoping to achieve?

Reply:
Removed comment prefix function. Yeah it was just for this function and I like `null_writes_queued` much more